src/components/Experience.ts

```diff
@@ -29,11 +30,85 @@ export class Experience<TGameStartr extends FullScreenPokemon> extends GeneralCo
      *
      * @param pokemon   The pokemon to give experience to.
      * @param experience   The amount of experience to give.
+     * @returns Whether the Pokemon leveled up.
      */
-    public gainExperience(pokemon: IPokemon, experience: number): void {
+    public gainExperience(pokemon: IPokemon, experience: number): boolean {
         pokemon.experience += experience;
         if (pokemon.experience >= this.gameStarter.equations.experienceStarting(pokemon.title, pokemon.level + 1)) {
             this.levelup(pokemon);
+            return true;
+        }
+        return false;
+    }
+
+     /**
+      * Processes experience gain for one or more pokemon
+      *
+      * @param battleInfo   Information about the current battle.
+      * @param onComplete   Handler for when this is done.
+      */
+    public processBattleExperience(battleInfo: IBattleInfo, onComplete: () => void): void {
+        const experienceToGain =  this.gameStarter.equations.experienceGained(
+            battleInfo.teams.player, battleInfo.teams.opponent);
+
+        const experienceText: (string | string[])[][] = [[
+            battleInfo.teams.player.selectedActor.nickname,
+            ` gained ${experienceToGain} experience!`]];
+        this.gameStarter.menuGrapher.createMenu("GeneralText");
+        const levelUp = this.gainExperience(battleInfo.teams.player.selectedActor, experienceToGain);
+        if (levelUp) {
+            this.gameStarter.menuGrapher.addMenuDialog(
+                "GeneralText", experienceText,
+                () => this.processBattleLevelUp(battleInfo.teams.player.selectedActor, onComplete));
+        } else {
+            this.gameStarter.menuGrapher.addMenuDialog("GeneralText", experienceText, onComplete);
```

Comment:
Duplicated code of `this.gameStarter.menuGrapher.addMenuDialog("GeneralText", experienceText,`. Instead of having that same section of code twice, make a local variable equal to a different callback (`onComplete` or the level-up processing that eventually calls `onComplete`).